etc/afpd/afpstats.c

```diff
@@ -22,31 +22,74 @@
 #include <stdlib.h>
 #include <string.h>
 
-#include <dbus/dbus-glib.h>
 #include <glib.h>
+#include <gio/gio.h>
 
 #include <atalk/compat.h>
 #include <atalk/errchk.h>
 #include <atalk/logger.h>
 #include <atalk/server_child.h>
 
 #include "afpstats_obj.h"
-#include "afpstats_service_glue.h"
+//#include "afpstats_service_glue.h"
 
 /*
  * Beware: this struct is accessed and modified from the main thread
  * and from this thread, thus be careful to lock and unlock the mutex.
  */
 static server_child_t *childs;
+static GDBusNodeInfo *introspection_data = NULL;
+
+static void handle_method_call (GDBusConnection *connection,
+                    const gchar           *sender,
+                    const gchar           *object_path,
+                    const gchar           *interface_name,
+                    const gchar           *method_name,
+                    GVariant              *parameters,
+                    GDBusMethodInvocation *invocation,
+                    gpointer               user_data)
+{
+}
+
+static GVariant *handle_get_property (GDBusConnection  *connection,
+                     const gchar      *sender,
+                     const gchar      *object_path,
+                     const gchar      *interface_name,
+                     const gchar      *property_name,
+                     GError          **error,
+                     gpointer          user_data)
+{
+}
+
+
+static gboolean handle_set_property (GDBusConnection  *connection,
+                     const gchar      *sender,
+                     const gchar      *object_path,
+                     const gchar      *interface_name,
+                     const gchar      *property_name,
+                     GVariant         *value,
+                     GError          **error,
+                     gpointer          user_data)
+{
+}
+
+static const GDBusInterfaceVTable interface_vtable =
+{
+  handle_method_call,
+  handle_get_property,
+  handle_set_property,
```

Comment:
If your interface doesn't have any properties, you can (and should) leave the get-property and set-property callbacks set to `NULL`. You don't need to provide a stub implementation.

If you do provide a stub implementation, it needs to set the error and return `NULL` or `FALSE` as appropriate. A minimal implementation would be something like:

```
static GVariant handle_get_property (..., GError **error, gpointer user_data)
{
  g_set_error (error, G_DBUS_ERROR, G_DBUS_ERROR_NOT_SUPPORTED, "Not implemented");
  return NULL;
}

static gboolean handle_set_property (..., GError **error, gpointer user_data)
{
  g_set_error (error, G_DBUS_ERROR, G_DBUS_ERROR_NOT_SUPPORTED, "Not implemented");
  return FALSE;
}
```

But, again, setting both `get_property` and `set_property` to NULL is easier.

Reply:
The only use for this interface right now is a one-way blanket request for server stats. I went ahead and NULLed out the callbacks.